File: java/leetcode_3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        int left = 0;
        int right = 0;
        int len = s.length();
        Set<Character> set = new HashSet<>();
        
        while(left < len && right < len) {
            if( set.contains(s.charAt(right)) == false ) {
                set.add(s.charAt(right++));
                ans = Math.max(ans, right - left);
            } else {
                set.remove(s.charAt(left++));
            }
        }
        
        return ans;
        
    }
}

File: swift/leet_code_3.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var left = 0
        var right = 0
        var strLength = s.length
        var charSet = Set<Character>()
        var ret = 0
        
        while left < strLength && right < strLength {
            let currentChar = s[s.index(s.startIndex, offsetBy: right)]
            //print("\(strLength):\(left)-\(right) currentChar:\(currentChar)")
            if charSet.contains(currentChar) == false {
                charSet.insert(currentChar)
                right += 1
                if ret < right - left {
                    ret = right - left
                }
            } else {
                let toRemoveChar = s[s.index(s.startIndex, offsetBy: left)]
                left += 1
                charSet.remove(toRemoveChar)
            }
        }
        
        return ret
    }
}
